File: words.py
'''
read in the csv file of words,transcriptions into a hashmap*

*in Python 3.8 and later, hashmap key order is
order of key insertion. written in Python 3.10.12
'''
import csv

csv_path = './words.csv'

# hashmap with key = word, value = transcription (strings)
wordMap = {}
wordsInOrder = []   # in case Will wants to keep the words in order

with open(csv_path, mode='r', encoding='utf-8') as file:
    csv_reader = csv.reader(file)

    # Skip the header row if your CSV has one
    next(csv_reader, None)

    # Populate the dictionary
    for row in csv_reader:
        word = row[0]  # Assuming the word is in the first column
        transcription = row[1]  # Assuming the transcription is in the second column
        wordMap[word] = transcription
        wordsInOrder.append(word)

print(wordMap)

'''
return the next word and its transcription in a tuple
'''
wordsAsKeys = list(wordMap.keys())    # in insertion order Python 3.10

print(wordsAsKeys, " ", type(wordsAsKeys))

wordIndex = 0

def getNextWord():
    global wordIndex

    # if we've reached the last word
    if wordIndex >= len(wordsAsKeys):
        wordIndex = 0   # loop back around to first

    word = wordsAsKeys[wordIndex]
    transcription = wordMap[word]

    wordIndex += 1
    return (word, transcription)

print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')
print(getNextWord(),'\n')

# next: implement random functionality

File: public/words.js
const fs = require('fs');
const csv = require('csv-parser');

const csvPath = './words.csv';
const wordMap = {};
const wordsInOrder = [];

// read in the csv (words.csv)
// words.csv should have two columns, Word and Transcription
// callback function runs when readCSV complete.
(function readCSV() {
  console.log("Reading words.csv");
  fs.createReadStream(csvPath)
    .pipe(csv())
    .on('data', (row) => {
      const word = row['Word']; // Replace with your actual column header
      const transcription = row['Transcription']; // Replace with your actual column header
      wordMap[word] = transcription;
      wordsInOrder.push(word);
    });
})();


let wordIndex = 0;
function getNextWord() {
    if (wordIndex >= wordsInOrder.length) {
        wordIndex = 0; // Loop back around to first
    }
    const word = wordsInOrder[wordIndex];
    const transcription = wordMap[word];
    wordIndex++;
    return [word, transcription]; // Return as a tuple equivalent (array)
}

module.exports = getNextWord;
